python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py

```diff
@@ -279,9 +291,39 @@ def _chat_message_content_to_dict(self, message: "ChatMessageContent") -> dict[s
 
     async def _send_chat_request(self, settings: OpenAIChatPromptExecutionSettings) -> list["ChatMessageContent"]:
         """Send the chat request."""
-        response = await self._send_request(request_settings=settings)
+        span = start_completion_activity(
+            settings.ai_model_id,
+            self.MODEL_PROVIDER_NAME,
+            self._settings_messages_to_prompt(settings.messages),
+            settings,
+        )
+
+        try:
+            response = await self._send_request(request_settings=settings)
+        except Exception as exception:
+            set_completion_error(span, exception)
+            span.end()
+            raise
+
         response_metadata = self._get_metadata_from_chat_response(response)
-        return [self._create_chat_message_content(response, choice, response_metadata) for choice in response.choices]
+
+        chat_message_contents = [
+            self._create_chat_message_content(response, choice, response_metadata) for choice in response.choices
+        ]
+
+        if span:
+            finish_reasons: list[str] = [choice.finish_reason for choice in response.choices]
+            with trace.use_span(span, end_on_exit=True):
+                set_completion_response(
+                    span,
+                    chat_message_contents,
+                    finish_reasons,
+                    response.id,
+                    response.usage.prompt_tokens,
+                    response.usage.completion_tokens,
+                )
+
+        return chat_message_contents
```

Comment:
`span.end()` is only called in the error case, after `set_completion_error` is called. Questions:

- Should `span.end()` be called in the success case as well?
- Is there ever a need to use the span after calling `set_completion_x` but before calling `end`? If not, should the span be ended in those operations i.e. `end_span_with_error`

Reply:
So span.end() marks the time at which a span ends.

Notice at line 313 the following code:
`with trace.use_span(span, end_on_exit=True):`

**end_on_exit=True** means that when we exit, the context manager takes care of ending the span.

This means that we do end the span on successful calls (under the hood) and that we could use the span after set_completion if we wanted to.